### PR TITLE
Write version API RFC

### PR DESCRIPTION
**Pull request by malt3**
### Proposed change(s)

- Write version API RFC

### Additional info

- This is a really early draft idea. I am not proposing it at this point in time.
- The document exists as a PR so it is easier to comment on.
- Not every aspect has to be implemented right away

### Checklist

- [ ] Update [CHANGELOG.md](https://github.com/edgelesssys/constellation/blob/main/CHANGELOG.md)
- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [ ] Link to Milestone

### CODE REVIEW

**Review by datosh — COMMENTED** [review]
Great idea. I think this is both flexible and simple enough for our use case.

One thing that came to mind is that our microservices already use Helm, which is the only system that already provides some sort of update discovery.

```sh
$ helm search repo nginx-ingress --version 1.10 -l
NAME                    CHART VERSION   APP VERSION     DESCRIPTION                                       
stable/nginx-ingress    1.10.3          0.25.0          An nginx Ingress controller that uses ConfigMap...
stable/nginx-ingress    1.10.2          0.25.0          An nginx Ingress controller that uses ConfigMap...
stable/nginx-ingress    1.10.1          0.25.0          An nginx Ingress controller that uses ConfigMap...
stable/nginx-ingress    1.10.0          0.25.0          An nginx Ingress controller that uses ConfigMap...

helm search repo nginx-ingress --version 1 -l
NAME                    CHART VERSION   APP VERSION     DESCRIPTION                                       
stable/nginx-ingress    1.41.3          v0.34.1         DEPRECATED! An nginx Ingress controller that us...
stable/nginx-ingress    1.41.2          v0.34.1         An nginx Ingress controller that uses ConfigMap...
stable/nginx-ingress    1.41.1          v0.34.1         An nginx Ingress controller that uses ConfigMap...
[... lots more ...]
```

Would it make sense to use this functionality? Either by dropping microservices from this RFC or by using the information helm provides as a backend to generate this information?

**Comment by 3u13r** [comment]
> Would it make sense to use this functionality? Either by dropping microservices from this RFC or by using the information helm provides as a backend to generate this information?

Since the helm charts are bundled as part of the CLI, we must upgrade it in order to upgrade the helm deployments. Also, we currently don't host them in a helm repository which is searchable.

File: rfc/update-api.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    - Kubernetes versions
+
+The folowing HTTP endpoints are available:
+
+- `GET /constellation/v1/updates/latest/` contains files showing the latest version available
+    - `image.json` contains the latest image version
+    - `microservice.json` contains the latest microservice version
+    - `cli.json` contains the latest cli version
+    - `kubernetes.json` contains the latest supported version of Kubernetes
+- `GET /constellation/v1/updates/major/<major-version>/` contains files with version information for this major version
```
Comment (3u13r) [inline rfc/update-api.md:22]: I'm thinking about the tradeoff between defining different endpoints and just having one endpoint and do the version parsing client side. I think that one endpoint with all version is more generic, which benefits third parties. 

Also defining one endpoint and doing client side parsing allows for more flexibility.

File: rfc/update-api.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    - `image.json` contains a list of all minor image versions that belong to a major version
+    - `microservice.json` contains a list of all minor microservice versions that belong to a major version
+    - `cli.json` contains a list of all minor cli versions that belong to a major version
+    - `kubernetes.json` contains a list of all supported minor version of Kubernetes that belong to a major version
+- `GET /constellation/v1/updates/minor/<minor-version>/` contains files with version information for this minor version
+    - `image.json` contains a list of all patch image versions that belong to a minor version
+    - `microservice.json` contains a list of all patch microservice versions that belong to a minor version
+    - `cli.json` contains a list of all patch cli versions that belong to a minor version
+    - `kubernetes.json` contains a list of all supported patch version of Kubernetes that belong to a minor version
```
Comment (3u13r) [inline rfc/update-api.md:31]: Should we combine some of those? I.e. I can only obtain a new microservice version through updating the CLI. So it is more of a meta information about the specific CLI version. I think we should at least somehow reflect which microservice version and Kubernetes version a specific CLI supports.

Comment (malt3) [reply rfc/update-api.md:31]: I think we should not impose version constraints early on in this API. Keeping them separate allows us to change compatibility checks between resources more easily in the CLI.

File: rfc/update-api.md (hunk 1 above)
Comment (malt3) [reply rfc/update-api.md:22]: > Also defining one endpoint and doing client side parsing allows for more flexibility.

Multiple endpoints give us more flexibility for each resource type.
We could add metadata like changelogs, download URLs or other info that is unique to one kind of resource.

**Comment by malt3** [comment]
@katexochen would like to name this API `version`(s) API. I think that's a good idea as it describes clearly what the API does / doesn't do.

**Pull request merged**